docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,622 @@
+    }
+
+    interface Builder {
+        /**
+         * The custom S3AsyncClient this transfer manager will use to make calls
+         * to S3.
+         */
+        Builder s3client(S3AsyncClient s3Client);
+
+        /**
+         * The configuration for this transfer manager.
+         */
+        Builder configuration(TransferManagerConfiguration configuration);
+
+        /**
+         * Add a progress listener to the currently configured list of
+         * listeners.
+         */
+        Builder addProgressListener(TransferProgressListener progressListener);
+
+        /**
+         * Set the list of progress listeners, overwriting any currently
+         * configured list.
+         */
+        Builder progressListeners(Collection<? extends TransferProgressListener> progressListeners);
+
+        S3TransferManager build();
+    }
+}
+
+/**
+ * Configuration object for {@link S3TransferManager}.
+ */
+public interface TransferManagerConfiguration {
+    /**
+     * The aggregate max upload rate in bytes per second over all active upload
+     * transfers. The default value is {@link Long#MAX_VALUE}, or effectively
+     * unlimited.
+     */
+    long maximumUploadBytesPersSecond();
+
+    /**
+     * The aggregate max download rate in bytes per second over all active
+     * download transfers. The default value is {@link Long#MAX_VALUE}, or
+     * effectively unlimited.
+     */
+    long maximumDownloadBytesSecond();
+
+    /**
+     * Whether multipart uploads are enabled.
+     */
+    boolean enableMultipartUpload();
+
+    /**
+     * The minimum length that object must be in order for it to be uploaded
+     * using the MultiPart API. This value must be at least 5 MiB, the minimum size
+     * allowed by S3 for an object part (excluding the last part).
+     */
+    long multipartUploadThreshold();
+
+    /**
+     * The minimum size for each part of an multipart upload.
+     */
+    int minimumUploadPartSize();
+
+    /**
+     * Whether multipart downloads are enabled.
+     */
+    boolean enableMultipartDownload();
+
+    /**
+     * The maximum number of parts an object be will downloaded in.
+     */
+    long maximumDownloadPartCount();
+
+    /**
+     * The minimum size for each part of a multipart download.
+     */
+    long minimumDownloadPartSize();
+}
+
+/**
+ * Override configuration for a single transfer.
+ */
+public interface TransferOverrideConfiguration {
+    /**
+     * The maximum rate for this transfer in bytes per second.
+     */
+    long maxTransferBytesPerSecond();
+
+    /**
+     * Whether or not multipart object downloads/uploads are enabled for this
+     * transfer.
+     */
+    boolean enableMultipartTransfer();
+}
+
+/**
+ * A factory capable of creating the streams for individual parts of a given
+ * object to be uploaded to S3.
+ * <p>
+ * There is no ordering guaranatee for when {@link
+ * #streamForPart(PartUploadContext)} is called.
+ */
+public interface ObjectPartStreamCreator {
```

Comment:
The name seems a bit confusing to me. If I understand it correctly, this is essentially a factory for `AsyncRequestBody`?